src/libostree/ostree-repo-composefs.c

```diff
@@ -0,0 +1,631 @@
+
+  /* Cache any directory metadata we read during this operation;
+   * see commit b7afe91e21143d7abb0adde440683a52712aa246
+   */
+  g_auto (OstreeRepoMemoryCacheRef) memcache_ref;
+  _ostree_repo_memory_cache_ref_init (&memcache_ref, self);
+
+  g_assert_cmpint (g_file_info_get_file_type (source_info), ==, G_FILE_TYPE_DIRECTORY);
+
+  const char *dirtree_checksum = ostree_repo_file_tree_get_contents_checksum (source);
+  const char *dirmeta_checksum = ostree_repo_file_tree_get_metadata_checksum (source);
+  return checkout_composefs_recurse (self, dirtree_checksum, dirmeta_checksum, target->dest, "root",
+                                     cancellable, error);
+}
+
+static struct lcfs_node_s *
+ensure_lcfs_dir (struct lcfs_node_s *parent, const char *name, GError **error)
+{
+  struct lcfs_node_s *node;
+
+  node = lcfs_node_lookup_child (parent, name);
+  if (node != NULL)
+    return node;
+
+  node = lcfs_node_new ();
+  lcfs_node_set_mode (node, 0755 | S_IFDIR);
+  if (lcfs_node_add_child (parent, node, name) != 0)
+    {
+      lcfs_node_unref (node);
+      glnx_throw_errno (error);
+      return NULL;
+    }
+
+  return node;
+}
+#endif
+
+/**
+ * ostree_repo_checkout_composefs:
+ * @self: Repo
+ * @target: A target for the checkout
+ * @source: Source tree
+ * @cancellable: Cancellable
+ * @error: Error
+ *
+ * Check out @source into @target, which is an in-memory
+ * representation of a composefs image. The @target can be reused
+ * multiple times to layer multiple checkouts before writing out the
+ * image to disk using ostree_composefs_target_write().
+ *
+ * There are various options specified by @options that affect
+ * how the image is created.
+ *
+ * Returns: %TRUE on success, %FALSE on failure
+ */
+gboolean
+ostree_repo_checkout_composefs (OstreeRepo *self, OstreeComposefsTarget *target,
+                                OstreeRepoFile *source, GCancellable *cancellable, GError **error)
+{
+#ifdef HAVE_COMPOSEFS
+  char *root_dirs[] = { "usr", "etc", "boot", "var", "sysroot" };
+  int i;
+  struct lcfs_node_s *root, *dir;
+
+  g_autoptr (GFileInfo) target_info
+      = g_file_query_info (G_FILE (source), OSTREE_GIO_FAST_QUERYINFO,
+                           G_FILE_QUERY_INFO_NOFOLLOW_SYMLINKS, cancellable, error);
+  if (!target_info)
+    return FALSE;
+
+  if (!checkout_composefs_tree (self, target, source, target_info, cancellable, error))
+    return FALSE;
+
+  /* We need a root dir */
+  root = ensure_lcfs_dir (target->dest, "root", error);
+  if (root == NULL)
+    return FALSE;
+
+  /* To work as a rootfs we need some root directories to use as bind-mounts */
+  for (i = 0; i < G_N_ELEMENTS (root_dirs); i++)
+    {
+      dir = ensure_lcfs_dir (root, root_dirs[i], error);
+      if (dir == NULL)
+        return FALSE;
+    }
+
+  return TRUE;
+#else
+  g_set_error (error, G_IO_ERROR, G_IO_ERROR_NOT_SUPPORTED,
+               "Composeefs is not supported in this ostree build");
```

Comment:
s/composeefs/composefs/

(Also this error is duplicated in a few places, let's add a single helper function to set the error for this case)